### PR TITLE
Require virtual for unimplemented functions.

### PR DESCRIPTION
**Pull request by ekpyron**
This is the first step into what I think will become a series of override/virtual cleanup PRs...
I tried to make a lot of the test-cases non-abstract where appropriate (although not all of them for which it was possible).
Someone should have a careful look to check, if I got the intention of all tests right.

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -433,6 +433,8 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
 		m_errorReporter.typeError(_function.location(), "Constructor must be implemented if declared.");
 	else if (isLibraryFunction)
 		m_errorReporter.typeError(_function.location(), "Library functions must be implemented if declared.");
+	else if (!_function.virtualSemantics())
+		m_errorReporter.typeError(_function.location(), "Functions without implementation must be marked virtual.");
```
Comment (leonardoalt) [inline libsolidity/analysis/TypeChecker.cpp:437]: Why is it the case here that the function doesn't have an implementation?

**Review by leonardoalt — CHANGES_REQUESTED** [review]
Does it need a Changelog entry?

Please update the documentation tests, they are failing.

**Comment by chriseth** [comment]
No changelog needed, it is part of the virtual/override feature.
I will take over and merge.

File: libsolidity/analysis/TypeChecker.cpp (hunk 1 above)
Comment (ekpyron) [reply libsolidity/analysis/TypeChecker.cpp:437]: It's among the ``else`` cases of ``if (function.isImplemented())`` line 430.

**Comment by ekpyron** [comment]
Updated docs. solcjs tests still need fixing.

**Comment by ekpyron** [comment]
> solcjs tests still need fixing.

Might work now, let's see.

**Review by chriseth — APPROVED** [review]

**Pull request merged**